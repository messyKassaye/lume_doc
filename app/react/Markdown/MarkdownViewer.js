import PropTypes from 'prop-types';
import React, { Component } from 'react';
import rison from 'rison-node';

import CustomComponents from './components';
import CustomHookComponents from './CustomHooks';

import markdownToReact from './markdownToReact';

class MarkdownViewer extends Component {
  static errorHtml(index, message) {
    return (
      <p key={index} className="error">
        <br />
        <strong>
          <i>Custom component markup error: unsuported values! Please check your configuration</i>
        </strong>
        <br />
        {message}
        <br />
      </p>
    );
  }

  static customHook(config, index) {
    const props = rison.decode(config);
    if (!CustomHookComponents[props.component]) {
      throw new Error('Invalid  component');
    }
    const Element = CustomHookComponents[props.component];
    return <Element {...props} key={index} />;
  }

  inlineComponent(type, config, index) {
    const { compact } = this.props;
    let result;
    if (type === 'list') {
      result = this.list(config, index);
    }

    if (type === 'link') {
      result = <CustomComponents.MarkdownLink {...rison.decode(config)} key={index} />;
    }

    if (type === 'searchbox') {
      result = <CustomComponents.SearchBox {...rison.decode(config)} key={index} />;
    }

    if (['vimeo', 'youtube', 'media'].includes(type)) {
      result = <CustomComponents.MarkdownMedia key={index} config={config} compact={compact} />;
    }

    if (type === 'customhook') {
      result = MarkdownViewer.customHook(config, index);
    }
    return result;
  }

  customComponent(type, config, index, children) {
    try {
      if (typeof type === 'function') {
        const Element = type;
        return (
          <Element {...config} key={index}>
            {children}
          </Element>
        );
      }

      if (type) {
        return this.inlineComponent(type, config, index);
      }
    } catch (error) {
      return MarkdownViewer.errorHtml(index, error.message);
    }

    return false;
  }

  list(_config, index) {
    const listData = this.props.lists[this.renderedLists] || {};
    const output = (
      <CustomComponents.ItemList
        key={index}
        link={`/library/${listData.params}`}
        items={listData.items}
        options={listData.options}
      />
    );
    this.renderedLists += 1;
    return output;
  }

  render() {
    this.renderedLists = 0;

    const ReactFromMarkdown = markdownToReact(
      this.props.markdown,
      this.customComponent.bind(this),
      this.props.html
    );

    if (!ReactFromMarkdown) {
      return false;
    }

    return <div className="markdown-viewer">{ReactFromMarkdown}</div>;
  }
}

MarkdownViewer.defaultProps = {
  lists: [],
  markdown: '',
  html: false,
  compact: false,
};

MarkdownViewer.propTypes = {
  markdown: PropTypes.string,
  lists: PropTypes.arrayOf(PropTypes.object),
  html: PropTypes.bool,
  compact: PropTypes.bool,
};

export default MarkdownViewer;
