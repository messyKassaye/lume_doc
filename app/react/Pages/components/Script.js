import { Component } from 'react';
import PropTypes from 'prop-types';

class Script extends Component {
  componentDidMount() {
    this.appendScript();
  }

  componentDidUpdate(prevProps) {
    const { children } = this.props;
    if (children !== prevProps.children) {
      this.appendScript();
    }
  }

  appendScript() {
    const { children } = this.props;
    if (children) {
      const s = document.createElement('script');
      s.src = `data:text/javascript,(function(){${encodeURIComponent(`\n\n${children}\n\n`)}})()`;
      document.body.appendChild(s);
    }
  }

  render() {
    return null;
  }
}

Script.defaultProps = {
  children: '',
};

Script.propTypes = {
  children: PropTypes.string,
};

export default Script;
