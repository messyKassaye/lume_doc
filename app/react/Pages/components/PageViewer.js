import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import Footer from 'app/App/Footer';
import MarkdownViewer from 'app/Markdown';
import MDComponents from 'app/Markdown/components';

import Script from './Script';

const { Context } = MDComponents;

export class PageViewer extends Component {
  render() {
    const { page, itemLists, datasets } = this.props;
    const lists = itemLists.toJS();
    const originalText = page.getIn(['metadata', 'content']) || '';
    const scriptCode = page.getIn(['metadata', 'script']) || '';

    return (
      <div className="row">
        <Helmet title={page.get('title') ? page.get('title') : 'Page'} />
        <main className="page-viewer document-viewer">
          <div className="main-wrapper">
            <Context.Provider value={datasets}>
              <MarkdownViewer html markdown={originalText} lists={lists} />
            </Context.Provider>
            <Footer />
          </div>
        </main>
        <Script>{scriptCode}</Script>
      </div>
    );
  }
}

PageViewer.defaultProps = {
  page: Immutable.fromJS({}),
  itemLists: Immutable.fromJS([]),
  datasets: Immutable.fromJS({}),
};

PageViewer.propTypes = {
  page: PropTypes.instanceOf(Immutable.Map),
  itemLists: PropTypes.instanceOf(Immutable.List),
  datasets: PropTypes.instanceOf(Immutable.Map),
};

const mapStateToProps = ({ page }) => ({
  page: page.pageView,
  datasets: page.datasets,
  itemLists: page.itemLists,
});

export default connect(mapStateToProps)(PageViewer);
