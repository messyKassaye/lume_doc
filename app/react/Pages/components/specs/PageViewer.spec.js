/** @format */

import { fromJS as Immutable } from 'immutable';
import React from 'react';

import { shallow } from 'enzyme';
import MarkdownViewer from 'app/Markdown';

import { PageViewer } from '../PageViewer';
import Script from '../Script';

describe('PageViewer', () => {
  let component;
  let props;
  let context;

  beforeEach(() => {
    props = {
      page: Immutable({
        _id: 1,
        title: 'Page 1',
        metadata: /*non-metadata-object*/ { content: 'MarkdownContent', script: 'JSScript' },
      }),
      itemLists: Immutable([{ item: 'item' }]),
    };
  });

  const render = () => {
    component = shallow(<PageViewer {...props} />, { context });
  };

  describe('render', () => {
    beforeEach(() => {
      render();
    });

    it('should render a MarkdownViewer with the markdown and the items for the lists', () => {
      expect(component.find(MarkdownViewer).props().markdown).toBe('MarkdownContent');
      expect(component.find(MarkdownViewer).props().lists).toEqual([{ item: 'item' }]);
    });

    it('should render the script', () => {
      const scriptElement = component.find(Script);
      expect(scriptElement).toMatchSnapshot();
    });
  });
});
