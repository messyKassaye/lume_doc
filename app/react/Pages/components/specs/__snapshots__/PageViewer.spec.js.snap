// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageViewer render should render the script 1`] = `
<Script>
  JSScript
</Script>
`;
